Add Total Particle & Field Momentum to Reduced Diagnostics

EZoni opened the pull request:
This PR adds the total particle and field momentum to the reduced diagnostics, which might be useful when testing the conservation properties of the different algorithms implemented in WarpX.

Since the field momentum requires the computation of cross products between MultiFabs with potentially different staggerings, it is necessary to use cell-centered MultiFabs to perform that computation.

I also used the new tuple-level reductions for particle containers available in AMReX (see https://github.com/ECP-WarpX/WarpX/pull/1908#discussion_r638241148), for both particle momentum and particle energy. Hence, some extra clean-up of ParticleEnergy.cpp is included here, too.

EZoni commented on Source/Diagnostics/ReducedDiags/FieldMomentum.cpp:
```diff
@@ -0,0 +1,147 @@
+        const amrex::MultiFab & By = warpx.getBfield(lev, 1);
+        const amrex::MultiFab & Bz = warpx.getBfield(lev, 2);
+
+        // Allocate cell-centered MultiFabs
+        amrex::MultiFab Ex_cc(amrex::convert(Ex.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Ey_cc(amrex::convert(Ey.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Ez_cc(amrex::convert(Ez.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Bx_cc(amrex::convert(Bx.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab By_cc(amrex::convert(By.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        amrex::MultiFab Bz_cc(amrex::convert(Bz.boxArray(), amrex::IntVect::TheZeroVector()),
+                              warpx.DistributionMap(lev), 1, 0);
+
+        // Interpolate to cell-centered MultiFabs
+        CoarsenIO::Coarsen(Ex_cc, Ex, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Ey_cc, Ey, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Ez_cc, Ez, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Bx_cc, Bx, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(By_cc, By, 0, 0, 1, 0);
+        CoarsenIO::Coarsen(Bz_cc, Bz, 0, 0, 1, 0);
```
In Source/Diagnostics/ReducedDiags/FieldMaximum.cpp we interpolate to cell-centers on the fly, by calling directly `CoarsenIO::Interp` (instead of `CoarsenIO::Coarsen`) while looping over cells. However, here the sum/loop over cells is hidden inside the function `amrex::MultiFab::Dot`, implemented in [amrex/Src/Base/AMReX_MultiFab.cpp](https://github.com/AMReX-Codes/amrex/blob/a61b40464f3f16eff420d19d99cff0fa010eb2b6/Src/Base/AMReX_MultiFab.cpp#L35-L80). So I'm not sure how I could avoid the allocation of the cell-centered MultiFabs in this case.

EZoni replied:
One possibility would probably be to do it as in #1944 (Source/Diagnostics/ReducedDiags/FieldReduction.H). However, since the logic hidden inside `amrex::MultiFab::Dot` seems quite complicated, I'm just not 100% sure that the two solutions would be fully equivalent.

ax3l replied:
maybe @atmyers has an idea? :)

atmyers replied:
I believe it's equivalent - the logic in `amrex::MultiFab::Dot` is complicated mostly because it has to handle CPU versus GPU execution; the actual dot product part is simple. In the case of `Source/Diagnostics/ReducedDiags/FieldReduction.H`, the CPU/GPU part is handled in the `reduce_op`, so I think the approach there (calling `CoarsenIO::Interp` with the proper staggering flags inside the lambda passed to the reduce_op` should work.

ax3l commented on Source/Diagnostics/ReducedDiags/FieldMomentum.cpp:
```diff
@@ -0,0 +1,147 @@
+#include "Utils/WarpXConst.H"
+#include "Utils/CoarsenIO.H"
+
+#include <AMReX_REAL.H>
```
Some missing includes to increase robustness:
```suggestion
#include <AMReX.H>
#include <AMReX_Geometry.H>
#include <AMReX_MultiFab.H>
#include <AMReX_ParallelDescriptor.H>
#include <AMReX_ParmParse.H>
#include <AMReX_REAL.H>
```

ax3l commented on Source/Diagnostics/ReducedDiags/FieldMomentum.cpp:
```diff
@@ -0,0 +1,147 @@
+#include <AMReX_REAL.H>
+
+#include <iostream>
+#include <cmath>
```
`iostream` is a really big header, I think we only need `ostream` here for `std::ofstream` & `std::endl`:
```suggestion
#include <cmath>
#include <ostream>
#include <string>
```
```

ax3l commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.H:
```diff
@@ -0,0 +1,39 @@
+#define WARPX_DIAGNOSTICS_REDUCEDDIAGS_PARTICLEMOMENTUM_H_
+
+#include "ReducedDiags.H"
+#include <fstream>
```
```suggestion
#include <string>
```

ax3l commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp:
```diff
@@ -0,0 +1,260 @@
+#include <AMReX_REAL.H>
+#include <AMReX_ParticleReduce.H>
+
+#include <iostream>
```
```suggestion
#include <ostream>
```

ax3l commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp:
```diff
@@ -0,0 +1,260 @@
+
+#include <iostream>
+#include <cmath>
+#include <limits>
```
```suggestion
#include <limits>
#include <string>
#include <vector>
```

ax3l reviewed (commented):
Great idea!

Please also update the user-facing `.rst` files for the manual :)

EZoni commented:
Updates:
- https://github.com/ECP-WarpX/WarpX/pull/1908/commits/9b8e8d473ad459beb449f8ef01039208fdffa4a5: fixed `includes`
- https://github.com/ECP-WarpX/WarpX/pull/1908/commits/9fb0fb4e0a69f31fc147bbb55ee89cd2da7c6917: added documentation of new reduced diagnostics for field and particle momentum
(also improved some formulas and descriptions for field and particle energy, to be consistent)
- https://github.com/ECP-WarpX/WarpX/pull/1908/commits/3feba24865dbf738fa84f4456305180a51a1634f: added interpolation to cell centers on the fly
(thanks to @NeilZaim for hinting at how to do this and thanks to @ax3l and @atmyers for confirming that it was indeed possible to use the `amrex::Reduce` tools)

ax3l commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp:
```diff
@@ -0,0 +1,262 @@
+        amrex::Real Py = 0.0_rt;
+        amrex::Real Pz = 0.0_rt;
+
+        Px = ReduceSum(myspc, [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
```
I think we can pass multiple lambdas here, but am not sure if this creates only one kernel.

I would avoid iterating the particle species multiple times, because that is generally a costly operation. Can you try to
create a single lambda that returns `Real[4]` for the three components and sum of `w`?

EZoni replied:
Yes, thank you for the suggestion. I'll mark the PR as "[WIP]" again and try to see if I can do that.

EZoni replied:
@ax3l As discussed on Slack, this cannot be done currently, but @atmyers plans to add support for it in the future (update: added in https://github.com/AMReX-Codes/amrex/pull/2082).

EZoni replied:
I used the new tuple-level reductions for particle containers in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/fbec288743f2bcc5b5edb2d65c3bb780d9dac86b, so now we have a single lambda that returns all necessary quantities.

NeilZaim commented on Source/Diagnostics/ReducedDiags/FieldMomentum.H:
```diff
@@ -0,0 +1,37 @@
+#define WARPX_DIAGNOSTICS_REDUCEDDIAGS_FIELDMOMENTUM_H_
+
+#include "ReducedDiags.H"
+#include <fstream>
```
```suggestion
```

Don't think you need that include. And if we want to be very exhaustive with includes, I guess you could add `#include <string>` instead (although being exhaustive with includes looks hard and tedious after looking at #1947 :grin: )

NeilZaim commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp:
```diff
@@ -0,0 +1,262 @@
+    // Get number of species
+    const int nSpecies = mypc.nSpecies();
+
+    // Get species names
+    const std::vector<std::string> species_names = mypc.GetSpeciesNames();
+
```
```suggestion
```

Don't think we need that here.

NeilZaim commented on Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp:
```diff
@@ -0,0 +1,262 @@
+        // 3 values of total momentum for each species +
+        // 3 values of mean  momentum for all  species +
+        // 3 values of mean  momentum for each species
+        offset_mean_species = 3 + nSpecies*3 + 3 + i_s*3;
+        m_data[offset_mean_all+0] += m_data[offset_mean_species+0];
+        m_data[offset_mean_all+1] += m_data[offset_mean_species+1];
+        m_data[offset_mean_all+2] += m_data[offset_mean_species+2];
```
I don't think the mean momentum for all species should be the sum of the mean momentum of individual species. (if I have two species that have the same mean momentum `p_mean`, I think the mean momentum for all species should be `p_mean`, not `2*p_mean`)
It's probably better to divide the total momentum for all species by the sum of weights over all species.
I think the same should be done for the FieldEnergy reduced diag as well.

EZoni replied:
About this one, if you have two species both with mean momentum `p_mean` and you want what we call "mean momentum for all species" to be `p_mean` as well (which makes sense, I agree with you), shouldn't we divide simply by the number of species then? So `2*p_mean / 2 = p_mean` in the example. I don't think it would be guaranteed to give `p_mean` if we divide by the sum of weights over all species, would it? I could be missing something.

NeilZaim replied:
> I don't think it would be guaranteed to give `p_mean` if we divide by the sum of weights over all species, would it?

Yes, we would get `p_mean` if we divide the _total_ momentum (i.e. what is in the first 3 columns of this diag) by the sum of all weights of both species.

When you're writing `2*p_mean / 2 = p_mean`, it looks like you're suggesting to sum the _average_ momentum of each species (and then divide by the number of species). But I don't think we should do that because that would give the same weight to species that only contain a few (or even 0) particles as to species that contain many particles.

I'm not sure how much this value makes sense anyways but I feel like it's better to output the average momentum of all the particles in the simulation than the average of the mean momentum of each species. In the latter case, simply adding a species with no particle would change the average momentum, which I don't think makes sense.

EZoni replied:
Sounds good, I think I implemented what you suggested (total momentum divided by sum of weights over all species) in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/226a96cdcf2e7a79d0350f70e328ecb70b493f6e. Let me know if this looks correct to you. If so, I will possibly change what we do for the particle energy accordingly as well.

NeilZaim replied:
Yes, that looks good to me, thanks! Good idea to add a check in the analysis script! And I also agree that we should also make the change for the particle energy diag.

EZoni replied:
Particle energy done in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/94836cc028755584423d9c412ff38061badbe1af. Interestingly, there was some numerical cancellation issue causing a relative error of about 1e-7 (instead of machine precision) on the protons particle energy (which wasn't tested before). The modification
```cpp
-  return ( std::sqrt(us*c2 + c2*c2) - c2 ) * m * w;
+  return (std::sqrt(us * m2 * c2 + m2 * c4) - m * c2) * w;
```
should give a result that matches the one obtained by post-processing the yt data to machine precision. Not 100% sure about which one of the two is supposed to be more correct, from a numerical cancellation point of view, though. At least, I haven't looked into this in detail.

NeilZaim replied:
Yes it's good to know that the particle energy calculation is inaccurate in the nonrelativistic limit, thanks. (although it's not surprising when looking at the formula)

If we ever need to be more accurate in this regime, we could check if the kinetic energy is much smaller than the rest mass and if so use the classical formula `p^2/2m` (or even go a bit further in the Taylor expansion). I don't think this is really an issue for now though, and even a relative error of about 1e-7 is sufficient for most purposes.

Regarding which of your 2 formulas is the more correct, I think they should have the exact same accuracy since the problematic subtraction is the same (the terms are just multiplied by `m` in the second case). Of course the second formula will give you a machine precision error in the python analysis scripts, since the exact same one is used there, but I don't think it is more accurate per se. (actually it looks like the second formula does an extra multiplication per particle, but I don't think this should have any significant effect on the simulation speed, so we probably don't have to worry about that)

More generally, your PR looks good, thanks a lot! I just had a minor comment regarding the case where the total particle weight is 0. (see below)

EZoni replied:
Sounds good. The fact that the two formulas should have the same accuracy, because they just differ uniformly by a multiplication, is indeed why I was a but surprised to see that they lead to a 7-digit difference in the results. I agree that it is natural to match the Python calculation to machine precision if we use the exact same formula, but I would not expect to lose/gain 7 digits by using an alternative formula that supposedly has the same accuracy (because the form of the cancellation is the same).

Anyways, thanks for the suggestion on the total particle weight, I'll add a check on that. Then I'll wait until the latest AMReX commit becomes available in WarpX, in order to try using the new tuple-level reductions for particle containers that Andrew implemented, as Axel had suggested initially.

NeilZaim replied:
>  I would not expect to lose/gain 7 digits by using an alternative formula that supposedly has the same accuracy

I don't think we actually lose or gain accuracy by switching from one formula to the other. Both formulas yield results that are off by ~1e-7 (relatively) with respect to the actual mathematical result. And because the numerical error is mostly random, the two results that the formulas yield are also off with each other by ~1e-7 (which is what you saw in the first analysis script).

NeilZaim reviewed (commented):
Thanks a lot for this PR. Looks good to me, I only have minor comments.
You may also have to fix a few conflicts in the example input file and its analysis script now that the `FieldReduction` diag has been merged.

NeilZaim commented on Source/Diagnostics/ReducedDiags/ParticleEnergy.cpp:
```diff
@@ -133,43 +137,47 @@ void ParticleEnergy::ComputeDiags (int step)
+    // Total mean energy. Offset:
+    // 1 value of total energy for all  species +
+    // 1 value of total energy for each species
+    m_data[1+nSpecies] = m_data[0] / Wtot;
```
Maybe we should also check here if `Wtot` is 0 (same as in lines 160-163) to avoid getting `NaNs` in the simulation? That can also be done for the momentum.

EZoni replied:
This is a good idea. Done in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/04acc1736c79172d46792419e85c658c30cd57c7.

EZoni commented:
Note: I tested https://github.com/ECP-WarpX/WarpX/pull/1908/commits/fbec288743f2bcc5b5edb2d65c3bb780d9dac86b locally and it seemed to work. However, I think it will pass all automated checks on GitHub only after the next WarpX weekly update (when the newest AMReX commit will be fetched automatically).

NeilZaim reviewed (approved):
That looks good to me now. The new tuple reductions are useful!

ax3l commented on Source/Diagnostics/ReducedDiags/FieldMomentum.cpp:
```diff
@@ -0,0 +1,217 @@
+            const amrex::Array4<const amrex::Real> & Bz_arr = Bz[mfi].array();
+
+            // (E x B) along x
+            reduce_op_ExB_x.eval(box, reduce_data_ExB_x,
```
@EZoni @atmyers just as a potential follow-up PR: do you think we can use a single reduce for the fields as well? That would allow us to read the field components twice and spawn only one kernel.

ax3l replied:
Actually you already turn into a tuple, I guess that might already work?

EZoni replied:
Good idea, I'll try to see if I can do something similar to what I did for particles. I'll mark the PR as [WIP] until I see if that works.

ax3l replied:
Weiqun commented:
```C++
amrex::ReduceOps<ReduceOpSum, ReduceOpSum, ReduceOpSum> reduce_ops;
amrex::ReduceData<Real, Real, Real> reduce_data(reduce_ops);
for (MFIter ...) {
    reduce_ops.eval(box, reduce_data,
    [=] AMREX_GPU_DEVICE (int i, int j, int k) -> amrex::GpuTuple<Real, Real, Real>
    {
        const amrex::Real Ex_cc = ...;
        const amrex::Real Ey_cc = ...;
        const amrex::Real Ez_cc = ...;
        const amrex::Real Bx_cc = ...;
        const amrex::Real By_cc = ...;
        const amrex::Real Bz_cc = ...;
        return {Ey_cc * Bz_cc - Ez_cc * By_cc,
                Ez_cc * Bx_cc - Ex_cc * Bz_cc,
                Ex_cc * By_cc - Ey_cc * Bx_cc};
     });
}
auto r = reduce_data.value();
Real ExB_x = amrex::get<0>(r);
Real ExB_y = amrex::get<1>(r);
Real ExB_z = amrex::get<2>(r);
```

EZoni replied:
Thank you, done in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/4224952b3e34d2607f48eae901c73881929fabaf.

NeilZaim replied:
Hmm, looks like we could do this for the `FieldMaximum` (and potentially `RhoMaximum`) diag as well. I'll add this on my todo list.

ax3l reviewed (approved):
Looks great, just one inline question left about our field reduces on the same box.

ax3l reviewed (approved):
Awesome, thanks a lot! :rocket: :sparkles:
